TailLight/device.cpp

```diff
@@ -78,49 +216,24 @@ Routine Description:
         WDF_OBJECT_ATTRIBUTES attributes = {};
         WDF_OBJECT_ATTRIBUTES_INIT_CONTEXT_TYPE(&attributes, DEVICE_CONTEXT);
 
-        NTSTATUS status = WdfDeviceCreate(&DeviceInit, &attributes, &device);
+        status = WdfDeviceCreate(&DeviceInit, &attributes, &device);
         if (!NT_SUCCESS(status)) {
-            KdPrint(("TailLight: WdfDeviceCreate, Error %x\n", status));
+            KdPrint(("TailLight: WdfDeviceCreate, Error 0x%x\n", status));
             return status;
         }
     }
 
     // Driver Framework always zero initializes an objects context memory
-    DEVICE_CONTEXT* deviceContext = WdfObjectGet_DEVICE_CONTEXT(device);
-
-    {
-        // initialize DEVICE_CONTEXT struct with PdoName
-
-        // In order to send ioctls to our PDO, we have open to open it
-        // by name so that we have a valid filehandle (fileobject).
-        // When we send ioctls using the IoTarget, framework automatically 
-        // sets the filobject in the stack location.
-        WDF_OBJECT_ATTRIBUTES attributes = {};
-        WDF_OBJECT_ATTRIBUTES_INIT(&attributes);
-        attributes.ParentObject = device; // auto-delete with device
-
-        WDFMEMORY memory = 0;
-        NTSTATUS status = WdfDeviceAllocAndQueryProperty(device,
-            DevicePropertyPhysicalDeviceObjectName,
-            NonPagedPoolNx,
-            &attributes,
-            &memory);
+    DEVICE_CONTEXT* pDeviceContext = WdfObjectGet_DEVICE_CONTEXT(device);
 
-        if (!NT_SUCCESS(status)) {
-            KdPrint(("TailLight: WdfDeviceAllocAndQueryProperty failed 0x%x\n", status));
-            return STATUS_UNSUCCESSFUL;
-        }
-
-        // initialize pDeviceContext->PdoName based on memory
-        size_t bufferLength = 0;
-        deviceContext->PdoName.Buffer = (WCHAR*)WdfMemoryGetBuffer(memory, &bufferLength);
-        if (deviceContext->PdoName.Buffer == NULL)
-            return STATUS_UNSUCCESSFUL;
-
-        deviceContext->PdoName.MaximumLength = (USHORT)bufferLength;
-        deviceContext->PdoName.Length = (USHORT)bufferLength - sizeof(UNICODE_NULL);
+    status = AllocatePDOName(device, pDeviceContext->PdoName);
+    if (!NT_SUCCESS(status)) {
+        return status;
+    }
 
-        KdPrint(("TailLight: PdoName: %wZ\n", deviceContext->PdoName)); // outputs "\Device\00000083
+    status = AllocateHwIdString(device, pDeviceContext->OurHardwareId);
```

Comment:
I think you'll need to call `WdfDeviceGetIoTarget(target)` here to get the WDFIOTARGET from WDFDEVICE when calling `AllocateHwIdString`.